src/key/key.js

```diff
@@ -360,47 +359,77 @@ class Key {
+      const notNull = args.filter(arg => arg !== null);
+      if (notNull.length === 0) return null;
+      return Math.min(...notNull);
+    };
+
+    let primaryKeyExpiry;
+    try {
+      const { selfCertification } = await this.getPrimaryUser(null, userID, config);
+      const keyExpiry = helper.getKeyExpirationTime(this.keyPacket, selfCertification);
+      const sigExpiry = selfCertification.getExpirationTime();
+      // TODO check direct signatures
+      primaryKeyExpiry = toDateOrInfinity(Math.min(keyExpiry, sigExpiry));
+    } catch (e) {
+      primaryKeyExpiry = null;
+    }
+    if (!capabilities) return primaryKeyExpiry;
+
+    // loop through subkeys based on requested capabilities and retrieve their expiration times
+    const getLatestKeyExpiry = async keys => (keys.length === 0 ? null : toDateOrInfinity(
+      Math.max(...await Promise.all(
+        keys.map(async key => (key instanceof SubKey ? key.getExpirationTime(null, config) : primaryKeyExpiry))
+      ))
+    ));
+    let encryptExpiry = null;
     if (capabilities === 'encrypt' || capabilities === 'encrypt_sign') {
-      const encryptKey =
-        await this.getEncryptionKey(keyID, expiry, userID, { ...config, rejectPublicKeyAlgorithms: new Set(), minRSABits: 0 }).catch(() => {}) ||
-        await this.getEncryptionKey(keyID, null, userID, { ...config, rejectPublicKeyAlgorithms: new Set(), minRSABits: 0 }).catch(() => {});
-      if (!encryptKey) return null;
-      const encryptExpiry = await encryptKey.getExpirationTime(this.keyPacket, undefined, config);
-      if (encryptExpiry < expiry) expiry = encryptExpiry;
+      const encryptionKeys = (await Promise.all(
+        this.getKeys(keyID).map(key => this.getEncryptionKey(key.getKeyID(), null, userID, config).catch(() => null))
+      )).filter(Boolean);
+      if (encryptionKeys.length === 0) return null; // the key cannot encrypt
+      encryptExpiry = await getLatestKeyExpiry(encryptionKeys);
     }
+    let signExpiry = null;
     if (capabilities === 'sign' || capabilities === 'encrypt_sign') {
-      const signKey =
-        await this.getSigningKey(keyID, expiry, userID, { ...config, rejectPublicKeyAlgorithms: new Set(), minRSABits: 0 }).catch(() => {}) ||
-        await this.getSigningKey(keyID, null, userID, { ...config, rejectPublicKeyAlgorithms: new Set(), minRSABits: 0 }).catch(() => {});
-      if (!signKey) return null;
-      const signExpiry = await signKey.getExpirationTime(this.keyPacket, undefined, config);
-      if (signExpiry < expiry) expiry = signExpiry;
+      const signingKeys = (await Promise.all(
+        this.getKeys(keyID).map(key => this.getSigningKey(key.getKeyID(), null, userID, config).catch(() => null))
+      )).filter(Boolean);
+      if (signingKeys.length === 0) return null; // the key cannot sign
+      signExpiry = await getLatestKeyExpiry(signingKeys);
     }
-    return expiry;
+    return toDateOrInfinity(minNotNull(encryptExpiry, signExpiry));
```

Comment:
This is not quite sufficient, I think. Theoretically speaking, you could have a primary key for certification only which expires in 1 day, a signing subkey which expires in 2 days, and an encryption subkey which expires in 3 days. The overall expiry should still be in 1 day. That's kind of why the function was setup as it was, updating the expiry inside the ifs.

Reply:
Fixed in #1319